File: python/programmers/dfsbfs/p2.py
from collections import deque


class Network:
    def __init__(self, n, computers):
        self.n = n
        self.computers = computers
        self.graph = {i: set() for i in range(1, n + 1)}
        self.visit = [False for _ in range(n + 1)]

    def makeGraph(self):
        computers = self.computers
        for i in range(len(computers)):
            for j in range(len(computers[i])):
                if i != j and computers[i][j] == 1:
                    self.graph[i + 1].add(j + 1)
                    self.graph[j + 1].add(i + 1)
        return self.graph

    def bfs(self, start):
        queue = deque([start])
        while queue:
            out = queue.popleft()
            self.visit[out] = True
            for elm in self.graph[out]:
                if self.visit[elm] == False:
                    queue.append(elm)
                    self.visit[elm] = True

    def countNetwork(self):
        graph = self.makeGraph()
        ans = 0
        for i in range(1, len(graph) + 1):
            if self.visit[i] == False:
                ans += 1
                self.bfs(i)

        return ans


def solution(n, computers):
    network = Network(n, computers)
    ans = network.countNetwork()

    return ans

File: java/programmers/src/dfsbfs/p2.java
package dfsbfs;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;



public class p2 {

	static int n;
	static int[][] computers;
	static Set<Integer>[] graph;
	static boolean[] visit;

	public static void bfs(int start) {

		Queue<Integer> que = new LinkedList<>();

		que.add(start);

		while (!que.isEmpty()) {

			int out = que.poll();
			for (Integer elm : graph[out]) {
				if (visit[elm] == false) {
					visit[elm] = true;
					que.add(elm);

				}
			}

		}

	}

	public static void main(String[] args) {
		 n = 3;
		computers = new int[][] {{1, 1, 0}, {1, 1, 0}, {0, 0, 1}};

		graph = new HashSet[n+1];

		visit = new boolean[n+1];

		for (int i = 0; i < computers.length; i++) {
			graph[i] = new HashSet<>();

			for (int j = 0; j < computers[i].length; j++) {
				if (j != i && computers[i][j] == 1) {
					graph[i].add(j);
				}
			}

		}

		int ans = 0;
		for (int i = 0; i < n; i++) {
			if (visit[i] == false) {
				bfs(i);
				ans++;
			}
		}

		System.out.println(ans);

	}

}
